### PR TITLE
core_mmu: fix the ttb pa address setting

### PR DESCRIPTION
**Pull request by prime-zeng**
Using the real physic address to set the mmu ttbr, and don't rely on the
plat mapping.

Signed-off-by: Zeng Tao <prime.zeng@hisilicon.com>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
could we also fix the core_mmu_lpae.c.
in both core_init_mmu_regs()
`write_ttbr0_64bit((paddr_t)l1_xlation_table[get_core_pos()]);`
`write_ttbr0_el1((paddr_t)l1_xlation_table[get_core_pos()]);`

**Comment by jenswi-linaro** [comment]
I don't mind taking care of `core/arch/arm/mm/core_mmu_lpae.c` in a separate PR as there's more to fix that just that in that file.

**Comment by prime-zeng** [comment]
@jenswi-linaro i think etienne-lms is right, this PR should a complete fix.

**Comment by etienne-lms** [comment]
as far as using `virt_to_phys()`, i think an `assert(pa != 0)` could be nice, but that not mandatory.
otherwise, patch is ok.
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
I'm not done reviewing `core_mmu_lpae.c`

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 1]
```diff
@@ -495,6 +496,9 @@ void core_init_mmu_regs(void)
 {
 	uint32_t ttbcr = TTBCR_EAE;
 	uint32_t mair;
+	paddr_t ttbr0;
+
+	ttbr0 = virt_to_phys((void *)l1_xlation_table[get_core_pos()]);
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu_lpae.c]: Please be consistent with how you do this in the two versions of `core_init_mmu_regs()` in this file.

**Comment by jenswi-linaro** [comment]
`core_mmu_find_table()` in `core_mmu_lpae.c` need to be updated

**Comment by jenswi-linaro** [comment]
`core_mmu_create_user_map()` and `init_xlation_table()`  in `core_mmu_lpae.c` need to be updated

**Comment by prime-zeng** [comment]
It seems that  `core_mmu_find_table ,core_mmu_create_user_map and init_xlation_table`  don't need to be updated?

**Comment by jenswi-linaro** [comment]
Why is that? As far as I can tell they depend on flat mapping too.

**Comment by prime-zeng** [comment]
There is no where to operate the pa in those functions,  what i fix is to use the pa instead of va.

**Comment by etienne-lms** [comment]
I have been through the core_mmu/_lpae/v7/pager and friends. i think the only places where `virt_to_phys()` over level1 mmu tables were missing are those 3 patched here.

I have seen 3 other missing `virt_to_phys()` related to level2 (or level3) tables:
In core_mmu_lpae.c / `core_mmu_create_user_map()` :
```
-   map->user_map = (paddr_t)dir_info.table | TABLE_DESC;
+   map->user_map = virt_to_phys(dir_info.table) | TABLE_DESC;
```
In core_mmu_lpae.c / `init_xlation_table()` :
```
- 			desc = TABLE_DESC | (uint64_t)(uintptr_t)new_table;
+ 			desc = TABLE_DESC | virt_to_phys(new_table);
```
In core_mmu_v7.c / `map_page_memarea()` :
```
-        return (paddr_t)l2;
+        return virt_to_phys(l2);
```

**Comment by prime-zeng** [comment]
@etienne-lms @jenswi-linaro  thank you, you are right , i just only considered the level 1 translation, sorry about that.

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 2]
```diff
@@ -520,6 +524,9 @@ void core_init_mmu_regs(void)
 {
 	uint64_t mair;
 	uint64_t tcr;
+	paddr_t ttbr0;
+
+	ttbr0 = virt_to_phys((void *)l1_xlation_table[get_core_pos()]);
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu_lpae.c]: minor: i think the cast to `(void *)` is not necessary.

File: core/arch/arm/mm/core_mmu_lpae.c (hunk 1 above)
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu_lpae.c]: minor: i think the cast to `(void *)` is not necessary.

**Comment by etienne-lms** [comment]
2 minor comments, change look fine to me.

**Comment by prime-zeng** [comment]
Rebased and tags applied.

**Comment by jenswi-linaro** [comment]
This is a step in the right direction, but `core_mmu_find_table()` , `core_mmu_create_user_map()` and `init_xlation_table()` also depends on flat mapping. I'd prefer to have this fixed at the same time.

**Comment by prime-zeng** [comment]
@jenswi-linaro please help to check if anything missing?

**Comment by etienne-lms** [comment]
@prime-zeng i am ok with your change. yet i am not sure my proposal covered the whole xlat cases.
especially those pointed out by Jens that reads and uses xlat tables content, hence some physical addresses.

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 3]
```diff
@@ -396,7 +397,7 @@ static struct tee_mmap_region *init_xlation_table(struct tee_mmap_region *mm,
 				panic("running out of xlat tables");
 			memset(new_table, 0, XLAT_TABLE_SIZE);
 
-			desc = TABLE_DESC | (uint64_t)(uintptr_t)new_table;
+			desc = TABLE_DESC | virt_to_phys((void *)new_table);
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu_lpae.c]: cast to `void *` not needed.

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 4]
```diff
@@ -613,7 +620,9 @@ bool core_mmu_find_table(vaddr_t va, unsigned max_level,
 		/* Copy bits 39:12 from tbl[n] to ntbl */
 		ntbl = (tbl[n] & ((1ULL << 40) - 1)) & ~((1 << 12) - 1);
 
-		tbl = (uint64_t *)ntbl;
+		tbl = (uint64_t *)phys_to_virt((paddr_t)ntbl, MEM_AREA_TEE_RAM);
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu_lpae.c]: Cast to `uint64_t *` and `paddr_t` not needed.

File: core/arch/arm/mm/core_mmu_v7.c [hunk 5]
```diff
@@ -425,6 +424,10 @@ bool core_mmu_find_table(vaddr_t va, unsigned max_level,
 	} else {
 		uintptr_t ntbl = tbl[n] & ~((1 << 10) - 1);
 
+		ntbl = (uintptr_t)phys_to_virt((paddr_t)ntbl, MEM_AREA_TEE_RAM);
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu_v7.c]: Please use a helper variable of correct type to avoid all this casting.

**Comment by prime-zeng** [comment]
@jenswi-linaro   another question, i have read the source of map_page_memarea, i find that the unused entries has been zero filled, but i think maybe those entries has already been used, so it may bring problems.

**Comment by jenswi-linaro** [comment]
@prime-zeng, `map_page_memarea()` isn't perfect but it works for the first l2 table and since it's only used for one l2 table it's OK for the moment.

**Comment by jenswi-linaro** [comment]
By the way, it's quite hard to tell what's changed and not between the commits when force pushing each time. Our usual process is to push new commits addressing comments and when everything is OK squash and force push (and possibly rebase too). Right now I have to read through your entire patch and review it from scratch each time.

**Comment by jbech-linaro** [comment]
I agree, @prime-zeng , please try to just push differences to the same branch. In the end when we're all happy with everything, that's then we squash it all to a single patch (usually ...).

**Comment by prime-zeng** [comment]
@jenswi-linaro @jbech-linaro  agree, i think i have misunderstood the flow(last patch i was told to do the sqush, so i do it each time, now i know that when everything is ok we do the sqush), but i think it is hard to split into different patches now, so maybe need to review it from scratch(i think changes are few),sorry about that...

**Comment by jbech-linaro** [comment]
@prime-zeng , please go a head with the patches as planned. No big deal, but in the future, please try to just add diffs as we discussed. Thanks!

**Comment by prime-zeng** [comment]
@jbech-linaro  ok, one someone help review the latest one?   i think it is ok.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
Is this PR still relevant?

**Comment by etienne-lms** [comment]
i think it is and it should hit the mainline.
it seems some rebasing are required.

**Comment by jforissier** [comment]
OK thanks. I've rebased, resolved conflicts and merged => commit https://github.com/OP-TEE/optee_os/commit/e05c74a001b2cc50f5e551f7d9e45bfc29c6b9de.

**Pull request closed without merging**